book/numerical_methods/curve_fitting.nim

```diff
@@ -29,6 +31,253 @@ yes.
+    let beta = params[1]
+    let gamma = params[2]
+    let delta = params[3]
+    result = f(alpha, beta, gamma, delta, t)
+
+nbText: hlMd"""
+As we can see, all the parameters that we want to fit, are passed in as
+a single 1D Tensor that we unpack for clarity here. The only other thing
+that is needed is an initial guess of the parameters. We will set them to all 1 here: 
+"""
+
+nbCode:
+  let initialGuess = ones[float](4)
+
+nbText: "Now we are ready to do the actual fitting:"
+
+nbCodeInBlock:
+  let solution = levmarq(fitFunc, initialGuess, t, y)
+  echo solution
+
+nbText: hlMd"""
+As we can see, the found parameters are very close to the actual ones. But maybe we can do better,
+`levmarq` accepts an `options` parameter which is the same as the one described in [Optimization](./optimization.html)
+with the addition of the `lambda0` parameter. We can reduce the `tol` and see if we get an even better fit:
+"""
+
+nbCode:
+  let options = levmarqOptions(tol=1e-15)
+  let solution = levmarq(fitFunc, initialGuess, t, y, options=options)
+  echo solution
+
+nbText: hlMd"""
+As we can see, there isn't really any difference. So we can conclude that the
+found solution has in fact converged. 
+
+Here's a plot comparing the fitted and original function:
+"""
+
+block:
+  let ySol = tDense.map_inline:
+    fitFunc(solution, x)
+
+  var df = toDf({"t": tDense, "original": yDense, "fitted": ySol, "tPoint": t, "yPoint": y})
+  df = df.gather(@["original", "fitted"], key="Class", value = "y")
+  df.ggplot(aes("t", "y", color="Class")) +
+    geom_line() +
+    geom_point(aes("tPoint", "yPoint")) +
+    ggsave("images/levmarq_comparision.png")
+
+nbImage("images/levmarq_comparision.png")
+
+nbText: hlMd"""
+As we can see, the fitted curve is quite close to the original one.
+
+## Errors & Uncertainties
+We might also want to quantify exactly how good of a fit the computed weights give.
+One measure commonly used is [$\chi^2$](https://en.m.wikipedia.org/wiki/Pearson%27s_chi-squared_test):
+$$\chi^2 = \sum_i \frac{(y_i - \hat{y}(x_i))^2}{\sigma_i^2}$$ 
+where $y_i$ and $x_i$ are the measurements, $\hat{y}$ is the fitted curve and $\sigma_i$
+is the standard deviation (uncertainty/error) of each of the measurements. We will use the
+noise size we used when generating the samples here, but in general you will have to
+figure out yourself what errors are in your situation. It may for example be the resolution
+of your measurements or you may have to approximate it using the data itself. 
+
+We now create the error vector and sample the fitted curve:
+"""
+
+nbCode:
+  let yError = ones_like(y) * noise
+  let yCurve = t.map_inline:
+    fitFunc(solution, x)
+
+nbText: """
+Now we can calculate the $\chi^2$:
+"""
+
+nbCode:
+  var chi2 = 0.0
+  for i in 0 ..< y.len:
+    chi2 += ((y[i] - yCurve[i]) / yError[i]) ^ 2
+  echo "χ² = ", chi2
+
+nbText: hlMd"""
+Great! Now we have a measure of how good the fit is, but what if we add more points?
```

Comment:
Generally this goes under the terminology of "goodness-of-fit tests" (https://en.wikipedia.org/wiki/Goodness_of_fit). There's a huge variety of tests, depending on what one needs and in real world use cases it can be useful to look at multiple tests for one fit.

And further (as this can be a source of confusion), the χ² goodness of fit *test* is in principle completely independent of the least squares method. Effectively a simple least square method literally reduces χ². But one may use a χ² test after performing a least square fit or something else, or instead look at the χ² value after doing a maximum likelihood estimation. It's a good thing to keep in mind that the optimization routine and the goodness of fit tests are fully independent things. Often when these things are taught imo this independence is not explained well enough to students.

Reply:
Oh, that's quite the collection of test :o 

Haha yes it is a bit confusing when using χ² for both the lstq and testing the fit so this cleared up a bit of my confusions, thanks :D